vamb/h_loss.py

```diff
@@ -0,0 +1,820 @@
+        label_to_subtree_node = project_to_subtree[label_to_node]
+        self.eval_label_map = LabelMap(
+                to_node=label_to_subtree_node,
+                to_target=label_to_subtree_node,
+        )
+
+    def calc_loss(self, labels_in, labels_out, mu, logsigma):
+        ce_labels = self.loss_fn(labels_out[:, 1:], labels_in)
+
+        ce_labels_weight = 1. #TODO: figure out
+        kld = -0.5 * (1 + logsigma - mu.pow(2) - logsigma.exp()).sum(dim=1).mean()
+        kld_weight = 1 / (self.nlatent * self.beta)
+        loss = ce_labels*ce_labels_weight + kld * kld_weight
+
+        _, labels_in_indices = labels_in.max(dim=1)
+        with _torch.no_grad():
+            gt_node = self.eval_label_map.to_node[labels_in_indices.cpu()]
+            prob = self.pred_fn(labels_out[:, 1:])
+        prob = prob.cpu().numpy()
+        pred = _infer.argmax_with_confidence(self.specificity, prob, 0.5, self.not_trivial)
+        pred = _hier.truncate_given_lca(gt_node, pred, self.find_lca(gt_node, pred))
+
+        return loss, ce_labels, kld, _torch.tensor(_np.sum(gt_node == pred))
+
+    def trainmodel(
+        self,
+        dataloader: _DataLoader[tuple[Tensor]],
+        nepochs: int = 500,
+        lrate: float = 1e-3,
+        batchsteps: Optional[list[int]] = [25, 75, 150, 300],
+        logfile: Optional[IO[str]] = None,
+        modelfile: Union[None, str, Path, IO[bytes]] = None,
+    ):
+        if lrate < 0:
+            raise ValueError(f"Learning rate must be positive, not {lrate}")
+
+        if nepochs < 1:
+            raise ValueError("Minimum 1 epoch, not {nepochs}")
+
+        if batchsteps is None:
+            batchsteps_set: set[int] = set()
+        else:
+            # First collect to list in order to allow all element types, then check that
+            # they are integers
+            batchsteps = list(batchsteps)
+            if not all(isinstance(i, int) for i in batchsteps):
+                raise ValueError("All elements of batchsteps must be integers")
+            if max(batchsteps, default=0) >= nepochs:
+                raise ValueError("Max batchsteps must not equal or exceed nepochs")
+            last_batchsize = dataloader.batch_size * 2 ** len(batchsteps)
+            if len(dataloader.dataset) < last_batchsize:  # type: ignore
+                raise ValueError(
+                    f"Last batch size of {last_batchsize} exceeds dataset length "
+                    f"of {len(dataloader.dataset)}. "  # type: ignore
+                    "This means you have too few contigs left after filtering to train. "
+                    "It is not adviced to run Vamb with fewer than 10,000 sequences "
+                    "after filtering. "
+                    "Please check the Vamb log file to see where the sequences were "
+                    "filtered away, and verify BAM files has sensible content."
+                )
+            batchsteps_set = set(batchsteps)
```

Comment:
We might want to refactor this out to a function, so we can use it for VAE, AAE and the VAEVAE